File: 72-edit-distance/main.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        
        size_t m = word1.size();
        size_t n = word2.size();

        vector<vector<int>> dp(m + 1, vector<int>(n + 1));
        for (size_t i = 0; i <= m; i++){
            dp[i][0] = i;
        }
        for (size_t j = 0; j <= n; j++) {
            dp[0][j] = j;
        }

        for (size_t i = 1; i < m + 1; i++) {
            for (size_t j = 1; j < n + 1; j++){
                int cost = word1[i - 1] == word2[j - 1] ? 0 : 1;
                dp[i][j] = min({
                    dp[i - 1][j] + 1,
                    dp[i][j - 1] + 1,
                    dp[i - 1][j - 1] + cost,
                });
            }
        }

        return dp[m][n];
    }
};

File: 72-edit-distance/main.go
func minDistance(word1 string, word2 string) int {
    m, n := len(word1), len(word2)
    dp := buildDistanceMatrix(m, n)

    for idx := 1; idx < m + 1; idx++ {
        for jdx := 1; jdx < n + 1; jdx++ {
            cost := 1
            if word1[idx - 1] == word2[jdx - 1] {
                cost = 0
            }
            dp[idx][jdx] = min(
                dp[idx - 1][jdx] + 1,
                dp[idx][jdx - 1] + 1,
                dp[idx - 1][jdx - 1] + cost,
            )
        }
    }

    return dp[m][n]
}

func buildDistanceMatrix(m, n int) [][]int {
    dp := make([][]int, m + 1)

    for idx := range dp {
        dp[idx] = make([]int, n + 1)
    }

    for idx := 0; idx <= m; idx++ {
        dp[idx][0] = idx
    }

    for jdx := 0; jdx <= n; jdx++ {
        dp[0][jdx] = jdx
    }

    return dp
}

func min(nums ...int) int {
    _min := math.MaxInt
    for _, num := range nums {
        if num < _min {
            _min = num
        }
    }
    return _min
}
